### PR TITLE
Fixed #25746 -- Isolated inlined test models registration.

### PR DESCRIPTION
**Pull request by charettes**

### CODE REVIEW

File: django/test/utils.py [hunk 1]
```diff
@@ -640,3 +642,47 @@ def setUp(self):
...
+                self.disable()
+
+            decorated.setUp = setUp
+            decorated.tearDown = tearDown
+            return decorated
+        else:
+            @wraps(decorated)
+            def inner(*args, **kwargs):
+                with self:
+                    return decorated(*args, **kwargs)
```
Comment (charettes) [inline django/test/utils.py]: I was hesitant to pass the isolated registry as an argument here since this is not a pattern used by other test decorators.

File: tests/apps/tests.py [hunk 2]
```diff
@@ -275,14 +276,14 @@ class LazyA(models.Model):
 
         # Test models appearing twice, and models appearing consecutively
         model_keys = [('apps', model_name) for model_name in ['lazya', 'lazyb', 'lazyb', 'lazyc', 'lazya']]
-        apps.lazy_model_operation(test_func, *model_keys)
+        LazyA._meta.apps.lazy_model_operation(test_func, *model_keys)
```
Comment (charettes) [inline tests/apps/tests.py]: This is an example of a test that would benefit from being able to access `apps` as a passed argument.

File: django/contrib/contenttypes/checks.py [hunk 3]
```diff
@@ -1,17 +1,23 @@
...
 from __future__ import unicode_literals
 
+from itertools import chain
+
 from django.apps import apps
 from django.utils import six
 
 
-def check_generic_foreign_keys(**kwargs):
+def check_generic_foreign_keys(app_configs=None, **kwargs):
```
Comment (charettes) [inline django/contrib/contenttypes/checks.py]: Adjusted this method to correctly deal with specified `app_configs`. I guess this could be part of another commit.

File: django/core/checks/model_checks.py [hunk 4]
```diff
@@ -11,21 +12,24 @@
...
-                errors.append(
-                    Error(
-                        "The '%s.check()' class method is "
-                        "currently overridden by %r." % (
-                            model.__name__, model.check),
-                        hint=None,
-                        obj=model,
-                        id='models.E020'
-                    )
+    if app_configs is None:
```
Comment (charettes) [inline django/core/checks/model_checks.py]: Ditto.

File: django/contrib/auth/checks.py [hunk 5]
```diff
@@ -6,10 +6,19 @@
...
+    if app_configs is None:
+        cls = apps.get_model(settings.AUTH_USER_MODEL)
+    else:
+        app_label, model_name = settings.AUTH_USER_MODEL.split('.')
+        for app_config in app_configs:
+            if app_config.label == app_label:
+                cls = app_config.get_model(model_name)
+                break
+        else:
+            return []
```
Comment (charettes) [inline django/contrib/auth/checks.py]: Ditto.

**Comment by timgraham** [comment]
Pretty nice. Some documentation in the contributing guide about this problem and how the decorator solves it would be useful.

**Comment by timgraham** [comment]
I'd still like to see the contributing documentation mentioned above, but if you want to merge this so it doesn't go stale in the meantime, I don't mind if it comes later.

**Comment by charettes** [comment]
I'm planing to write docs today, I pushed some commits yesterday but couldn't finish.

**Comment by charettes** [comment]
@timgraham, moved docs to the contributing section and updated based on your comments.

File: docs/internals/contributing/writing-code/unit-tests.txt [hunk 6]
```diff
@@ -303,3 +303,114 @@ purpose.
...
+        class TestModelDefinition(SimpleTestCase):
+            @isolate_apps('app_label', 'other_app_label')
+            def test_model_definition(self):
+                class TestModel(models.Model):
+                    pass
+
+                class OtherAppModel(models.Model):
+                    class Meta:
+                        app_label = 'other_app_label'
+                ...
```
Comment (charettes) [inline docs/internals/contributing/writing-code/unit-tests.txt]: Is this what you had in mind @timgraham?

File: docs/internals/contributing/writing-code/unit-tests.txt [hunk 7]
```diff
@@ -303,3 +303,114 @@ purpose.
...
+    from django.test.utils import isolate_apps
+
+    class TestModelDefinition(SimpleTestCase):
+        @isolate_apps('app_label')
+        def test_model_definition(self):
+            class TestModel(models.Model):
+                pass
+            ...
+
+.. admonition:: app_label
```
Comment (timgraham) [inline docs/internals/contributing/writing-code/unit-tests.txt]: maybe:
```
Setting ``app_label``
```

File: docs/internals/contributing/writing-code/unit-tests.txt (hunk 6 above)
Comment (timgraham) [inline docs/internals/contributing/writing-code/unit-tests.txt]: Does this assume the test is in the app: `tests/app_label/tests.py`? You could use the snippet/filename directive to make it clear.

Comment (charettes) [inline docs/internals/contributing/writing-code/unit-tests.txt]: Done.

File: docs/internals/contributing/writing-code/unit-tests.txt [hunk 8]
```diff
@@ -303,3 +303,117 @@ purpose.
...
+unnecessary table creation, models defined in a test method should be bound to
+a temporary ``Apps`` instance::
+
+    from django.apps.registry import Apps
+    from django.db import models
+    from django.test import SimpleTestCase
+
+    class TestModelDefinition(SimpleTestCase):
+        def test_model_definition(self):
+            test_apps = Apps(['app_label])
```
Comment (timgraham) [inline docs/internals/contributing/writing-code/unit-tests.txt]: missing quote causing code highlighting to fail

File: docs/internals/contributing/writing-code/unit-tests.txt [hunk 9]
```diff
@@ -303,3 +303,117 @@ purpose.
...
+        @isolate_apps('app_label')
+        def test_model_definition(self):
+            class TestModel(models.Model):
+                pass
+            ...
+
+.. admonition:: Setting ``app_label``
+
+    Models defined in a test method with no explicit
+    :attr:`~django.db.models.Options.app_label` are also automatically
```
Comment (timgraham) [inline docs/internals/contributing/writing-code/unit-tests.txt]: chop "also"

File: docs/internals/contributing/writing-code/unit-tests.txt [hunk 10]
```diff
@@ -303,3 +303,117 @@ purpose.
...
+    assigned the label of the app in which their test class is located.
+
+    In order to make sure the models defined within the context of
+    :func:`~django.test.utils.isolate_apps` instances are correctly
+    installed you should pass the set of targeted ``app_label`` as arguments:
```
Comment (timgraham) [inline docs/internals/contributing/writing-code/unit-tests.txt]: installed,

File: docs/internals/contributing/writing-code/unit-tests.txt [hunk 11]
```diff
@@ -303,3 +303,117 @@ purpose.
...
+
+        from django.db import models
+        from django.test import SimpleTestCase
+        from django.test.utils import isolate_apps
+
+        class TestModelDefinition(SimpleTestCase):
+            @isolate_apps('app_label', 'other_app_label')
+            def test_model_definition(self):
+                class TestModel(models.Model):
+                    pass
```
Comment (timgraham) [inline docs/internals/contributing/writing-code/unit-tests.txt]: Add a comment `# This model automatically receives app_label='app_label'`

File: docs/internals/contributing/writing-code/unit-tests.txt [hunk 12]
```diff
@@ -303,3 +303,117 @@ purpose.
...
+    from django.test import SimpleTestCase
+
+    class TestModelDefinition(SimpleTestCase):
+        def test_model_definition(self):
+            test_apps = Apps(['app_label])
+
+            class TestModel(models.Model):
+                class Meta:
+                    apps = test_apps
+                    app_label = 'app_label'
```
Comment (timgraham) [inline docs/internals/contributing/writing-code/unit-tests.txt]: is the explicit app_label redundant here?

File: docs/internals/contributing/writing-code/unit-tests.txt [hunk 13]
```diff
@@ -303,3 +303,117 @@ purpose.
...
+            def test_model_definition(self):
+                class TestModel(models.Model):
+                    pass
+
+                class OtherAppModel(models.Model):
+                    class Meta:
+                        app_label = 'other_app_label'
+                ...
+
+The decorator can also be applied to :class:`unittest.TestCase` classes::
```
Comment (timgraham) [inline docs/internals/contributing/writing-code/unit-tests.txt]: chop ":class:`unittest.TestCase`"?

File: docs/internals/contributing/writing-code/unit-tests.txt (hunk 12 above)
Comment (charettes) [inline docs/internals/contributing/writing-code/unit-tests.txt]: It is.
